src/Strategy/Python/Poetry/Common.hs

```diff
@@ -1,65 +1,84 @@
 discover dir = context "Poetry" $ do
   projects <- context "Finding projects" $ findProjects dir
   pure (map mkProject projects)
 
+-- | Poetry build backend identifier required in [pyproject.toml](https://python-poetry.org/docs/pyproject/#poetry-and-pep-517).
+poetryBuildBackendIdentifier :: Text
+poetryBuildBackendIdentifier = "poetry.core.masonry.api"
+
+-- | Reference message text for poetry build backend setting value required in pyproject.toml.
+-- Users should configure poetry build backend in pyproject.toml for poetry project discovery.
+poetryBuildBackendIdentifierHelpText :: Text
+poetryBuildBackendIdentifierHelpText = "Poetry project must use poetry build backend. Please refer to https://python-poetry.org/docs/pyproject/#poetry-and-pep-517."
+
+-- | Finds poetry project by searching for pyrpoject.toml.
+-- If poetry.lock file is also discovered, it is used as a supplement.
 findProjects :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> m [PoetryProject]
 findProjects = walk' $ \dir _ files -> do
   let poetryLockFile = findFileNamed "poetry.lock" files
   let pyprojectFile = findFileNamed "pyproject.toml" files
 
   case (poetryLockFile, pyprojectFile) of
-    (Nothing, Just pyproject) -> do
+    (poetry, Just pyproject) -> do
       poetryProject <- readContentsToml pyProjectCodec pyproject
-      if usesPoetryBuildSystem poetryProject
-        then pure ([PoetryPyProjectOnly (ProjectDir dir) (PyProjectTomlFile pyproject)], WalkSkipAll)
-        else pure ([], WalkContinue)
-    (Just poetrylock, Just pyproject) ->
-      pure
-        (
-          [ PoetryPyProjectWithLock (ProjectDir dir) (PyProjectTomlFile pyproject) (PoetryLockFile poetrylock)
-          ]
-        , WalkSkipAll
-        )
-    -- Without pyproject file, it is unlikely that project is a poetry package. Poetry itself does not work
-    -- without pyproject.toml manifest.
+      let project = PoetryProject (ProjectDir dir) (PyProjectTomlFile pyproject) (PoetryLockFile <$> poetry)
+      let poetryBuildSystem = getPoetryBuildSystem poetryProject
+
+      case poetryBuildSystem of
+        Nothing -> pure ([], WalkContinue)
+        Just pbs ->
+          if pbs == poetryBuildBackendIdentifier
+            then pure ([project], WalkSkipAll)
+            else
+              context
+                ( "pyproject.toml does not use poetry build system. It uses: "
+                    <> pbs
+                    <> "\n"
+                    <> poetryBuildBackendIdentifierHelpText
+                )
+                $ pure ([], WalkContinue)
+
+    -- Without pyproject file, it is unlikely that project is a poetry project. Poetry itself does not work
+    -- without [pyproject.toml manifest](https://python-poetry.org/docs/pyproject/).
     (Just _, Nothing) -> context "poetry.lock file found without accompanying pyproject.toml!" $ pure ([], WalkContinue)
     _ -> pure ([], WalkContinue)
```

Comment:
If I'm correct, this can only catch the `(Nothing, Nothing)` case, since everything else is handled.  When possible, I like to avoid using `_` matches, since it lets the compiler tell us if we've handled every explicit case.

If you do make this change, and it doesn't compile (it may just warn locally, but in CI it will fail), it should tell us which cases we're not handling.  If it doesn't complain, then we've explicitly handled every case.

With `_`, it **always** compiles with no error/warning, which means we're weakening the compiler's guarantees.  It's unavoidable sometimes, and completely correct other times, but in cases like these, it's less helpful than it seems.

```suggestion
    (Nothing, Nothing) -> pure ([], WalkContinue)
```